lib/resty/session/memcached.lua

```diff
@@ -4,26 +4,135 @@
 -- @module resty.session.memcached
 
 
-local memcached = require "resty.memcached"
-local get_name = require "resty.session.utils".get_name
+local memcached   = require "resty.memcached"
+local utils       = require "resty.session.utils"
+local buffer      = require "string.buffer"
 
 
 local setmetatable = setmetatable
-local error = error
-local null = ngx.null
+local error        = error
+local null         = ngx.null
+local time         = ngx.time
+local get_name     = utils.get_name
+local get_meta_key = utils.get_meta_key
+local get_meta_el_val = utils.get_meta_el_val
+local get_latest_valid = utils.get_latest_valid
+
+
+local function metadata_cleanup(self, memc, aud_sub_key)
+  local now     = time()
+  local retry   = 10
+  local max_exp = now
+  local ok      = false
+
+  while(retry > 0 and not ok) do
+    retry = retry - 1
+    local res, _, cas_u, err = memc:gets(aud_sub_key)
+    if not res then
+      return nil, err
+    end
+
+    local sessions = get_latest_valid(res)
+    local buf  = buffer.new()
+
+    for s, exp in pairs(sessions) do
+      buf = buf:put(get_meta_el_val(s, exp))
+      max_exp = math.max(max_exp, exp)
+    end
+
+    ok, err = memc:cas(aud_sub_key, buf:tostring(), cas_u, max_exp - now)
+  end
+  return ok
+end
+
+local function read_metadata(self, memc, name, audience, subject)
+  local aud_sub_key = get_meta_key(self, audience, subject)
+  local res, _, err = memc:get(aud_sub_key)
+  if not res then
+    return nil, err
+  end
+
+  return get_latest_valid(res)
+end
+
+local function SET(self, memc, name, key, value, ttl, current_time, old_key, stale_ttl, metadata, remember)
+  local inferred_key = get_name(self, name, key)
+
+  if not metadata and not old_key then
+    return memc:set(inferred_key, value, ttl)
+  end
+
+  local ok, err = memc:set(inferred_key, value, ttl)
+  if err then
+    return nil, err
+  end
+
+  local old_name = old_key and get_name(self, name, old_key)
+  if old_name then
+    if remember then
+      memc:delete(old_name)
+    else
+      memc:touch(old_name, stale_ttl)
+    end
+  end
+
+  if metadata then
+    local audiences = metadata.audiences
+    local subjects  = metadata.subjects
+    for i = 1, #audiences do
+      local aud_sub_key = get_meta_key(self, audiences[i], subjects[i])
+      local meta_el_val = get_meta_el_val(key, current_time + ttl)
+      ok, err = memc:add(aud_sub_key, meta_el_val)
+      if not ok then
+        ok, err = memc:append(aud_sub_key, meta_el_val)
+      end
+      if old_key then
+        meta_el_val = get_meta_el_val(old_key, 0)
+        ok, err = memc:append(aud_sub_key, meta_el_val)
+      end
+      -- no need to clean up every time we write
+      -- it is just beneficial when a key is used a lot
+      if math.random() < 0.1 then
```

Comment:
cleaning up happens with a probability of 10% *on each aud_sub key*, when set() is called on it.

Reply:
Wouldn't it be simpler that it just happens on all or none, and not for `each`?

Reply:
2 reasons:

- we don't keep a list of all the audience:subject keys so with the current configuration it's easier to only clean up the "current" key that is being accessed
- cleaning up is more beneficial to keys that grow larger, so cleaning up each key individually based on a probability means keys that grow larger (are hit more times) have more chances of being cleaned up, so this should automatically balance size and clean ups of keys

Please if you can think of a better way let me know